File: mstparser/DependencyEvaluator.java
package mstparser;

import java.io.*;
import mstparser.io.*;

public class DependencyEvaluator {
	
    public static void evaluate(String act_file, String pred_file, String format) throws IOException {
	DependencyReader goldReader = DependencyReader.createDependencyReader(format);
	boolean labeled = goldReader.startReading(act_file);

	DependencyReader predictedReader = DependencyReader.createDependencyReader(format);
	boolean predLabeled = predictedReader.startReading(pred_file);

	if (labeled != predLabeled)
	    System.out.println("Gold file and predicted file appear to differ on whether or not they are labeled. Expect problems!!!");


	int total = 0; int corr = 0; int corrL = 0;
	int numsent = 0; int corrsent = 0; int corrsentL = 0;
	int root_act = 0; int root_guess = 0; int root_corr = 0;

	DependencyInstance goldInstance = goldReader.getNext();
	DependencyInstance predInstance = predictedReader.getNext();

	while(goldInstance != null) {

	    int instanceLength = goldInstance.length();

	    if (instanceLength != predInstance.length())
		System.out.println("Lengths do not match on sentence "+numsent);

	    int[] goldDeps = goldInstance.getDeps();
	    String[] goldLabels = goldInstance.get("labels");
	    int[] predDeps = predInstance.getDeps();
	    String[] predLabels = predInstance.get("labels");

	    boolean whole = true;
	    boolean wholeL = true;

	    // NOTE: the first item is the root info added during nextInstance(), so we skip it.

	    for (int i = 1; i < instanceLength; i++) {
		if (predDeps[i] == goldDeps[i]) {
		    corr++;
		    if (labeled) {
			if (goldLabels[i].equals(predLabels[i])) 
			    corrL++;
			else 
			    wholeL = false;
		    }
		}
		else { 
		    whole = false; wholeL = false; 
		}
	    }
	    total += instanceLength - 1; // Subtract one to not score fake root token

	    if(whole) corrsent++;
	    if(wholeL) corrsentL++;
	    numsent++;
						
	    goldInstance = goldReader.getNext();
	    predInstance = predictedReader.getNext();
	}

	System.out.println("Tokens: " + total);
	System.out.println("Correct: " + corr);
	System.out.println("Unlabeled Accuracy: " + ((double)corr/total));
	System.out.println("Unlabeled Complete Correct: " + ((double)corrsent/numsent));
	if(labeled) {
	    System.out.println("Labeled Accuracy: " + ((double)corrL/total));
	    System.out.println("Labeled Complete Correct: " + ((double)corrsentL/numsent));
	}
		
    }

    public static void main (String[] args) throws IOException {
	String format = "CONLL";
	if (args.length > 2)
	    format = args[2];

	evaluate(args[0], args[1], format);
    }

}

File: src/main/scala/mstparser/DependencyEvaluator.scala
package mstparser

import mstparser.io.DependencyReader

object DependencyEvaluator {
  def evaluate(actFile: String, predFile: String, format: String) = {
    val goldReader = DependencyReader.createDependencyReader(format)
    val labeled = goldReader.startReading(actFile)

    val predReader = DependencyReader.createDependencyReader(format)
    val predLabeled = predReader.startReading(predFile)

    if (labeled != predLabeled)
	    println("Gold file and predicted file appear to differ on whether or not they are labeled. Expect problems!!!")

    val (corr, corrL, corrSent, corrSentL) =
      goldReader.zip(predReader).zipWithIndex.foldLeft(0, 0, 0, 0) {
        case ((c, cL, cs, csL), ((goldInstance, predInstance), i)) =>
          goldInstance.difference(predInstance) match {
            case Some((d, dL)) => (
              c + goldInstance.size - d,
              cL + goldInstance.size - dL,
              cs + (if (d == 0) 1 else 0),
              csL + (if (dL == 0) 1 else 0)
            )
            case None =>
              println("Lengths do not match on sentence " + i)
              (c, cL, cs, csL)
	    }
    }

    println("Tokens: " + goldReader.tokenCount)
    println("Correct: " + corr)
    println("Unlabeled Accuracy: " + corr.toDouble / goldReader.tokenCount)
    println("Unlabeled Complete Correct: " + corrSent.toDouble / goldReader.instanceCount)
    if (labeled) {
      println("Labeled Accuracy: " + corrL.toDouble / goldReader.tokenCount)
      println("Labeled Complete Correct: " + corrSentL.toDouble / goldReader.instanceCount)
    }
  }

  def main(args: Array[String]) {
    val format = if (args.length > 2) args(2) else "CONLL"
    this.evaluate(args(0), args(1), format)
  }
}

